src/server/detail/save_stages_controller.h

```diff
@@ -56,14 +56,16 @@ class RdbSnapshot {
   }
 
   bool HasStarted() const {
-    return started_ || (saver_ && saver_->Mode() == SaveMode::SUMMARY);
+    return started_shards_.load(std::memory_order_relaxed) > 0 ||
```

Comment:
Is there an actual race here? `HasStarted()` is called in two places:

1. `SaveStagesController::SaveCb`
2. `SaveStagesController::GetSaveBuffersSize`

The first:

```
void SaveStagesController::WaitAllSnapshots() {                                                    
  RunStage(&SaveStagesController::SaveCb);                                                         
} 
```

Should be fine because by the time we return above `auto res = save_controller_->InitResourcesAndStart();` the `started_` member is properly initialized atomically (because we hold a lock on the `save_mu_`). 

For the second path it's pretty much the same thing:

```
lock_guard lk{save_mu_};                                                                    
  if (save_controller_) {                                                                     
    append("save_buffer_bytes", save_controller_->GetSaveBuffersSize());                      
}

```

What am I missing here?

Reply:
It doesn't matter how it's read. StartInShard runs concurrently, concurrent write to the same address is formally UB even it's harmless in this case

Reply:
but on a separate `snapshot` instance (disjoint, different address)
```
  // Start saving a dfs file on shard
  void SaveStagesController::SaveDfsSingle(EngineShard* shard) {
    // for summary file, shard=null and index=shard_set->size(), see SaveDfs() above
    auto& [snapshot, filename] = snapshots_[shard ? shard->shard_id() : shard_set->size()];
   // boilerplate code here deleted for simplciity
  
    if (auto err = snapshot->Start(mode, filename, glob_data); err) {
      shared_err_ = err;
      snapshot.reset();
      return;
    }
  
    if (mode == SaveMode::SINGLE_SHARD)
      snapshot->StartInShard(shard);
  }
```

So each shard modifies its own snapshot instance?

Reply:
Yes, but we have SaveRdb() 🙂

Reply:
wow that slipped right through my eyes. GOOD catch